kroxylicious-api/src/main/java/io/kroxylicious/proxy/config/tls/KeyPair.java

```diff
@@ -39,4 +57,76 @@ public record KeyPair(@JsonProperty(required = true) String privateKeyFile,
     public <T> T accept(KeyProviderVisitor<T> visitor) {
         return visitor.visit(this);
     }
+
+    /**
+     * Test if this certificate has SANs matching a set of hostnames
+     * @param hostnames hostnames
+     * @return true if all hostnames match an SAN of this certificate
+     */
+    public boolean matchesHostnames(Set<String> hostnames) {
+        Set<Pattern> patterns = hostnamePatterns();
+        return hostnames.stream().allMatch(hostname -> patterns.stream().map(pattern -> pattern.matcher(hostname)).anyMatch(Matcher::matches));
+    }
+
+    private Set<Pattern> hostnamePatterns() {
+        try {
+            Path certPath = Path.of(certificateFile);
+            String pemString = Files.readString(certPath);
+            X509Certificate certificate = parsePemX509(pemString);
+            Set<String> dnsSubjectAlternativeNames = getDnsSubjectAlternativeNames(certificate);
+            // fallback to common name
```

Comment:
I'm not expert here, but it is possible we could see a cert like CN=A and SANs=B,C,D.   Should hostnamePatterns be returning A,B,C,D?